underfs/gsg/src/main/java/alluxio/underfs/obs/GSGUnderFileSystem.java

```diff
@@ -0,0 +1,299 @@
+   * @return the created {@link GSGUnderFileSystem} instance
+   */
+  public static GSGUnderFileSystem createInstance(AlluxioURI uri, UnderFileSystemConfiguration conf)
+      throws IOException {
+    String bucketName = UnderFileSystemUtils.getBucketName(uri);
+    GoogleCredentials credentials;
+    if (conf.containsKey(GSGPropertyKey.GSG_CREDENTIAL_PATH)) {
+      LOG.info("Create GSGUnderFileSystem with UnderFileSystemConfiguration conf");
+      credentials = GoogleCredentials
+          .fromStream(new FileInputStream(conf.getValue(GSGPropertyKey.GSG_CREDENTIAL_PATH)))
+          .createScoped(Lists.newArrayList("https://www.googleapis.com/auth/cloud-platform"));
+    } else if (Configuration.isSet(GSGPropertyKey.GSG_CREDENTIAL_PATH)) {
+      LOG.info("Create GSGUnderFileSystem with Alluxio configuration");
+      credentials = GoogleCredentials
+          .fromStream(new FileInputStream(Configuration.get(GSGPropertyKey.GSG_CREDENTIAL_PATH)))
+          .createScoped(Lists.newArrayList("https://www.googleapis.com/auth/cloud-platform"));
+    } else {
+      LOG.info("Using Google application default credentials");
+      // The environment variable GOOGLE_APPLICATION_CREDENTIALS is set
+      credentials = GoogleCredentials.getApplicationDefault();
+    }
+
+    Storage storage = StorageOptions.newBuilder().setCredentials(credentials).build().getService();
+    Bucket bucket = storage.get(bucketName);
+    return new GSGUnderFileSystem(uri, storage, bucket, bucketName,
+        conf);
+  }
+
+  /**
+   * Constructor for {@link GSGUnderFileSystem}.
+   *
+   * @param uri the {@link AlluxioURI} for this UFS
+   * @param storageClient the Google cloud storage client
+   * @param bucketClient the Google cloud storage bucket client
+   * @param bucketName bucket name of user's configured Alluxio bucket
+   * @param conf configuration for this UFS
+   */
+  protected GSGUnderFileSystem(AlluxioURI uri, Storage storageClient, Bucket bucketClient,
+      String bucketName, UnderFileSystemConfiguration conf) {
+    super(uri, conf);
+    mStorageClient = storageClient;
+    mBucketClient = bucketClient;
+    mBucketName = bucketName;
+  }
+
+  @Override
+  public String getUnderFSType() {
+    return "GSG";
+  }
+
+  // Setting GSG owner via Alluxio is not supported yet. This is a no-op.
+  @Override
+  public void setOwner(String path, String user, String group) {}
+
+  // Setting GSG mode via Alluxio is not supported yet. This is a no-op.
+  @Override
+  public void setMode(String path, short mode) {}
+
+  @Override
+  protected boolean copyObject(String src, String dst) {
+    LOG.debug("Copying {} to {}", src, dst);
+    Storage.CopyRequest request = Storage.CopyRequest.newBuilder()
+        .setSource(BlobId.of(mBucketName, src))
+        .setTarget(BlobId.of(mBucketName, dst))
+        .build();
+    try {
+      Blob blob = mStorageClient.copy(request).getResult();
+      if (blob != null) {
+        return true;
+      }
```

Comment:
What does a `null` response without an exception mean? Should we log something or is it impossible?

Reply:
added the logging.
Google cloud api is strange, some times it will return null without exception means the file does not exist. More testing and investigation needed to be done on those client calls.